File: 2461-amount-of-time-for-binary-tree-to-be-infected/amount-of-time-for-binary-tree-to-be-infected.py
class Solution:
    def buildMap(self, node, parent_val, graph): #create undirected graph
        if not node:
            return
        if node.val not in graph:
            graph[node.val] = []
            if parent_val is not None:
                graph[node.val].append(parent_val)
                graph[parent_val].append(node.val)
            self.buildMap(node.left, node.val, graph)
            self.buildMap(node.right, node.val, graph)
    def amountOfTime(self, root: Optional[TreeNode], start: int) -> int:
        graph = defaultdict(list) 
        self.buildMap(root, None, graph)
        # Print the created graph
        print("Graph:")
        for node, neighbors in graph.items():
            print(f"{node}: {neighbors}")
        if start not in graph:
            print("start node not in graph")
            return -1
        vis = set([start])
        q = deque([start])
        time = 0
        while q:
            size = len(q)
            for i in range(size):
                node = q.popleft()
                for neighbor in graph[node]:
                    if neighbor in vis:
                        continue
                    vis.add(neighbor)
                    q.append(neighbor)
            time += 1
        return time-1
 

# class Solution: 		
#     def amountOfTime(self, root: Optional[TreeNode], start: int) -> int:
#         graph = defaultdict(list)
#         q = [(root, None)]
#         while q: 
#             n, p = q.pop()
#             if p: 
#                 graph[p.val].append(n.val)
#                 graph[n.val].append(p.val)
#             if n.left: q.append((n.left, n))
#             if n.right: q.append((n.right, n))
#         ans = -1
#         vis = {start}
#         q = deque([start])
#         while q: 
#             for _ in range(len(q)): 
#                 u = q.popleft()
#                 for v in graph[u]: 
#                     if v not in vis: 
#                         vis.add(v)
#                         q.append(v)
#             ans += 1
#         return ans 

File: 2461-amount-of-time-for-binary-tree-to-be-infected/amount-of-time-for-binary-tree-to-be-infected.cpp
class Solution {
public:
//build an undirected graph from the binary tree
    void makegraph(unordered_map<int,vector<int>> &adj,int parent,TreeNode* curr){
        if(curr==NULL) return;
        if(parent!=-1){//if curr node has a parent,connect it in graph
            adj[curr->val].push_back(parent);//store it in grpah
        }
        if(curr->left){//l child store
            adj[curr->val].push_back(curr->left->val);
        }if(curr->right){//r child
            adj[curr->val].push_back(curr->right->val);
        }
        makegraph(adj,curr->val,curr->left);
        makegraph(adj,curr->val,curr->right);
    }
    int amountOfTime(TreeNode* root, int start) {
        unordered_map<int,vector<int>> adj;//unord graph-adj list repr
 // Step 1- Convert the tree to graph
        makegraph(adj,-1,root);// -1=root has no parent
        queue<int> q;//q for bfs
        unordered_set<int> vis;//track infectd nodes
        q.push(start);//start frm infected ndoe
        vis.insert(start);//mark it 
        int min=0;//time
        while(!q.empty()){
            int n=q.size();//no of infected nodes at curr lvl-this sec
            while(n--){//goin over all nodes at curr lvl
                int curr=q.front();//curr infected node
                q.pop();
                for(int &neighbor:adj[curr]){//infect all unvis neigbors-child or parent
                    if(vis.find(neighbor)==vis.end()){
                        q.push(neighbor);//infect neigh
                        vis.insert(neighbor);//mark as vis
                    }
                }
            }
            min++;//1more sec/lvl passed
        }     
        return min-1;//since last lvl doesnt cause new infections,min-1
    }
};

//  //BFS
// class Solution {
// public:
//     unordered_map<int,vector<int>> v;  //adjacency list

//     //create undirected graph for every parent-child  -> BFS
//     void createGraph(TreeNode* root){
//         queue<pair<TreeNode*,int>> q;
//         q.push({root,-1});
//         while(q.size()){
//             auto [node,parent]= q.front(); q.pop();
//             if(parent!=-1){
//                 v[parent].push_back(node->val);
//                 v[node->val].push_back(parent);
//             }
//             if(node->left)  q.push({node->left,node->val});
//             if(node->right) q.push({node->right,node->val});
//         }   
//     }

//     int amountOfTime(TreeNode* root, int start) {
//         //create graph of given tree
//         createGraph(root);

//         //start bfs
//         queue<int> q;
//         unordered_map<int,bool> seen;
//         q.push(start);
//         seen[start]=true;
//         int time=0;
//         for(;q.size();time++){
//             int n= q.size();
//             while(n--){
//                 auto node= q.front();  q.pop();
//                 for(auto i:v[node]){
//                     if(!seen[i]){
//                         q.push(i);
//                         seen[i]=true;
//                     }
//                 }
//             }
//         }
//         return time-1;
//     }
// };
// // Time - O(n)
// // Space - O(n)